l3build.dtx

```diff
@@ -1246,6 +1248,38 @@
 % engine-specific, thus one |.tpf| file should be stored per engine to be
 % tested.
 %
+% \subsection{Custom tests}
+%
+% If neither the text-based methods nor PDF-based tests are sufficient,
+% there is the additional option of defining custom variants with individual
+% normalization rules.
+%
+% For this, the variant has to be registered in the \texttt{test_types} table
+% and then activated in \texttt{test_order}.
+%
+% Every element in \texttt{test_types} is a table with fields \texttt{test}
+% (the extension of the test file), \texttt{reference} (the extension of the
+% file the output is compared with), \texttt{generated} (extension of the
+% analyzed \LaTeX{} output file) and \texttt{rewrite} (A Lua function for
+% normalizing the output file, taking as parameters the name of the unnormalized
+% \LaTeX{} output file to be read, the name of the normalized file to be written,
+% the engine name and a potential errorcode).
+%
+% For example:
+% \begin{verbatim}
+% test_types = {
+%   mytest = {
+%     test = ".mylvt",
+%     reference = ".mytlg",
+%     generated = ".log",
+%     rewrite = function(source, normalized, engine, errorcode)
+%       -- In this example we just copy the logfile without any normalization
+%       os.execute(string.format("cp %s %s", source, normalized)
+%     end,
+%   },
+% }
```

Comment:
This does not match how everything else works. I'd expect `test_types  =  test_types or { }`, `test_types.mytest = test_types.mytext or { ... }`, the same as we do for 'special' engines and so on.

Reply:
What happens if only some entries are given? Do we need a prototype to be copied then altered, for example?

Reply:
I don't really understand what you mean here.

All the entries `test`, `reference` and `generated` are required because otherwise the system doesn't know which files to operate on. Similarly, `rewrite` is effectivly always required because there is no reasonable default (We could copy without normalization but that would never be reproducible.)

The case of copying and altering a prototype sounds like you would just have two parts in `rewrite`: First copy, then rewrite. (Or both steps can be combined as in the default normalization functions)